### PR TITLE
budget_managment/zainab1AT

### PR DESCRIPTION
**Pull request by zainab1AT**
## Implement Basic User and Budget Management Functionality

This pull request introduces a foundational structure for a user and budget management system. The system allows for the creation of user accounts, balance management through deposits and transfers, and retrieval of user information. 

### User Class:

* Attributes:

1. userName: Stores the unique username for each user.
2. firstName: Stores the first name of the user.
3. lastName: Stores the last name of the user.
4. balance: Initializes to 0 and stores the user's account balance.

* Methods:

1. setBalance(newBalance): Updates the user's balance.
2. getUsername(), getFirstname(), getLastname(), getBalance(): Retrieve the respective user information.

### BudgetManagementService Class:

* Attributes:

1. users: A Map object to store and manage users by their userName.

* Methods:

1. getUserByUsername(userName): Fetches user details if the username exists.
2. addUser(userName, firstName, lastName): Adds a new user if the username is not already taken.
3. deposit(username, amount): Allows depositing a positive amount into a user's account.
4. transfer(senderUsername, receiverUsername, amount): Facilitates transferring a valid amount from one user to another, with checks for sufficient balance.

### Command-Line Interface (CLI):

The app class provides a simple interface to interact with the BudgetManagementService, allowing users to:

- Add new users.
- Deposit money into user accounts.
- Transfer money between users.
- View user information.
- Exit the application.

## Enhancements and Considerations:

1. Error Handling & Input Validation: User-friendly error messages are provided to guide users when an invalid operation is attempted and basic validation to ensure that usernames are unique, deposit amounts are positive, and transfer amounts are valid.

## Testing:
Basic manual testing was conducted using the CLI to ensure the functionality works as expected.

### CODE REVIEW

**Comment by zainab1AT** [comment]
budget_managment/zainab1AT

**Comment by zainab1AT** [comment]
.

**Comment by mohammedabualsoud** [comment]
@zainab1AT Please look at the contribution guide, description and title of the pull request

File: node_modules/.package-lock.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (mohammedabualsoud) [inline node_modules/.package-lock.json]: You just need the package.json and package-lock.json
You should not push any external code(libraries/packages) to the repo.

You need to skip the node_modules/ directory from being tracked by git

File: projects/budget_managment/zainab1AT/User.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
+export default class User {
+  static users = new Map();
+  #userName;
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/User.js]: what are these #userName, #firstName ... ?
It's not common naming convention to start with #

File: projects/budget_managment/zainab1AT/User.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
+export default class User {
+  static users = new Map();
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/User.js]: How come a class of User contains list of Users ?
this is not a User it's something we Call it UserCollection/UserStore/UserService/UserDAL

File: projects/budget_managment/zainab1AT/User.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  getBalance() {
+    return this.#balance;
+  }
+
+  static getUserByUsername(userName) {
+    const user = User.users.get(userName);
+    if (user) {
+      return JSON.stringify({
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/User.js]: Why you are returning JSON string ?

File: projects/budget_managment/zainab1AT/User.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        username: User.users.get(userName).getUsername(),
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/User.js]: everyTime you all calling get user, to get a single info of the user.
You could do it once and use that user instance for getting all the details

File: projects/budget_managment/zainab1AT/User.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  deposit(amount) {
+    if (typeof amount === "number" && amount > 0) {
+      this.#balance += amount;
+    } else {
+      console.log("Invalid Deposition, enter valid value !");
+    }
+  }
+
+  transfer(receiverUsername, amount) {
+    if (typeof receiverUsername === "string" && typeof amount === "number") {
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/User.js]: there are nested if else statements here. You could make it more readable.

File: projects/budget_managment/zainab1AT/App.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    const choice = readlineSync.question('Please choose an option (1-5): ');
+
+    switch (choice) {
+        case '1':
+            console.clear();
+            console.log('--- Add User ---');
+            const username = readlineSync.question('Enter Username: ');
+            const fname = readlineSync.question('Enter First Name: ');
+            const lname = readlineSync.question('Enter Last Name: ');
+            new User(username, fname, lname);
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/App.js]: if the user already exists, your output is not correct. the user added User added successfully will be printed for the user.

File: projects/budget_managment/zainab1AT/User.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    } else {
+      console.log("This User doesn't exist");
+    }
+  }
+
+  setBalance(newBalance) {
+    this.#balance = newBalance;
+  }
+
+  deposit(amount) {
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/User.js]: following the Single Responsibility Principle, the User class is doing more than one job.
This is not something for the user, I think better to split the non-user class responsibility to another class maybe the Bank / BudgetManagerService/BudgetService that expose the deposit & transfer capabilities APIs

**Review by mohammedabualsoud — CHANGES_REQUESTED** [review]
Please check my comments, and add unit tests too.

File: projects/budget_managment/zainab1AT/User.js (hunk 2 above)
Comment (zainab1AT) [reply projects/budget_managment/zainab1AT/User.js]: I use # to make the fields private, ensuring they can't be accessed or modified directly from outside the class, protecting the internal state of my object. Especially since they are matters related to money, it's very private.

File: projects/budget_managment/zainab1AT/User.js (hunk 3 above)
Comment (zainab1AT) [reply projects/budget_managment/zainab1AT/User.js]: Does that mean I have to create another class and put it in?

File: projects/budget_managment/zainab1AT/User.js (hunk 4 above)
Comment (zainab1AT) [reply projects/budget_managment/zainab1AT/User.js]: When I see the form of the example that you put it in the readme file for this idea , It occurred to me that it is JSON.

File: projects/budget_managment/zainab1AT/App.js (hunk 7 above)
Comment (zainab1AT) [reply projects/budget_managment/zainab1AT/App.js]: Okay, I corrected it.

File: .gitignore
Comment (mohammedabualsoud) [inline .gitignore:1]: Please make sure to follow the **contribution guidelines**, all the files should be created within your **app** directory

File: projects/budget_managment/zainab1AT/README.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+* Exit: Choose option 5 to exit the application.
+
+# Installation
+
+* Clone the repository:
+``` git clone https://github.com/yourusername/budget-management-app.git ```
+
+* Install dependencies:
+Install the necessary packages by running:
+```npm install readline-sync```
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/README.md]: clearly the library here is dependecny of the app. So this should be part of the package.json or whatever tool you used to manage the dependencies.
in your case the user should run this command to install the dependencies

`npm install`

File: projects/budget_managment/zainab1AT/User.js (hunk 2 above)
Comment (mohammedabualsoud) [reply projects/budget_managment/zainab1AT/User.js]: Please read the airbnb javascript best practices guide.

https://github.com/airbnb/javascript?tab=readme-ov-file#naming--leading-underscore

File: projects/budget_managment/zainab1AT/User.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import BudgetManagementService from './BudgetManagementService.js'; 
+export default class User {
+  #userName;
+  #firstName;
+  #lastName;
+  #balance;
+
+  constructor(userName, firstName, lastName) {
+    if (BudgetManagementService.users.has(userName)) {
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/User.js]: well you have made the User & BudgetManagementService very coupled.
I think better to make the User class a simple data class representing the User, and the Budget one could be responsible to maintain the users. I saw that part are coupled between the two classes, which is not the best design

File: projects/budget_managment/zainab1AT/tempCodeRunnerFile.js
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/tempCodeRunnerFile.js:1]: I think not needed. Please remove

**Comment by mohammedabualsoud** [comment]
@zainab1AT Your Pull Request missing the unit tests & the description is empty. Could you please fill these 2 requirements

File: projects/budget_managment/zainab1AT/User.js (hunk 2 above)
Comment (mohammedabualsoud) [reply projects/budget_managment/zainab1AT/User.js]: @zainab1AT Can you apply the coding style I asked for please

File: projects/budget_managment/zainab1AT/User.js (hunk 4 above)
Comment (mohammedabualsoud) [reply projects/budget_managment/zainab1AT/User.js]: Better to not convert it to JSON string unless there's a reason which is not the case I guess

File: projects/budget_managment/zainab1AT/BudgetManagementService.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
+import User from "./User.js";
+export default class BudgetManagementService {
+  constructor() {
+    this.users = new Map();
+  }
+  getUserByUsername(userName) {
+    const user = this.users.get(userName);
+    if (user) return user;
+    else throw new Error("This User not found !");
```
Comment (mohammedabualsoud) [inline projects/budget_managment/zainab1AT/BudgetManagementService.js:9]: no need for else block
